orchestrator/careplanservice/service.go

```diff
@@ -674,3 +675,63 @@ func (s *Service) checkAllowUnmanagedOperations() error {
 	}
 	return nil
 }
+
+// validateLiteralReferences validates the literal references in the given resource.
+// Literal references may be an external URL, but they MUST use HTTPS and be a child of a FHIR base URL
+// registered in the CSD. This prevents unsafe external references (e.g. accidentally exchanging resources over HTTP),
+// and gives more confidence that the resource can safely be fetched by SCP-nodes.
+func (s *Service) validateLiteralReferences(ctx context.Context, resource any) error {
+	// Literal references are "reference" fields that contain a string. This can be anywhere in the resource,
+	// so we need to recursively search for them.
+	resourceAsJson, err := json.Marshal(resource)
+	if err != nil {
+		// would be very weird
+		return err
+	}
+	resourceAsMap := make(map[string]interface{})
+	err = json.Unmarshal(resourceAsJson, &resourceAsMap)
+	if err != nil {
+		// would be very weird
+		return err
+	}
+
+	literalReferences := make(map[string]string)
+	collectLiteralReferences(resourceAsMap, []string{}, literalReferences)
+	for path, reference := range literalReferences {
+		lowerCaseRef := strings.ToLower(reference)
+		if strings.HasPrefix(lowerCaseRef, "http://") {
+			return coolfhir.BadRequest(fmt.Sprintf("literal reference is URL with scheme http://, only https:// is allowed (path=%s)", path))
+		}
+		if strings.HasPrefix(lowerCaseRef, "https://") {
+			// OK
+			// TODO: Check if it's a child of a registered FHIR base URL
+			parsedRef, err := url.Parse(reference)
+			if err != nil {
+				// weird
+				return err
+			}
+			if slices.Contains(strings.Split(parsedRef.Path, "/"), "..") {
+				return coolfhir.BadRequest(fmt.Sprintf("literal reference is URL with parent path segment '..' (path=%s)", path))
```

Comment:
I wonder if Azure FHIR/HAPI would use the parameters... But, it does not seem to be disallowed:

> Absolute URLs do not need to point to a [FHIR RESTful server](https://build.fhir.org/http.html), though this is the preferred approach. Whether or not the reference is to a FHIR RESTful server, the reference SHALL point to a Resource as defined by this specification.

(https://build.fhir.org/references.html#literal)

I don't see any risk in allowing it?

Reply:
Azure/HAPI won't. But we need to prepare for "any" client injecting a URL that orca will try to resolve I would say. They are "trusted" parties, setting an external reference, but they might have an open vulnerability.
 
I didn't know FHIR allowed to params in the references, as they always need to point to a single resource. But they do not explicitly forbid it indeed. I was thinking, params might enable some cleverly crafted SSRF attack as we try to resolve the URL, where it would access the internal network (probably hard due to the http denial, but who knows). Probably a lot less to "worry" about when you only allow for "clean" links.

Reply:
Ok, then I'll add the check. We can always allow it again at some later point.